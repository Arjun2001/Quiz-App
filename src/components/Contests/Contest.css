#title {
    text-align: center;
    margin-top: 15px;
  }
  
  #employee {
    border-collapse: collapse;
    border: 3px solid #ddd;
  }
  
  #employee td,
  #employee th {
    border: 1px solid #ddd;
    padding: 12px;
  }
  
  #employee tr:hover {
    background-color: #ddd;
  }
  
  #employee th {
    padding: 10px;
    text-align: center;
    background-color: #4caf50;
    color: white;
  }
  
  .opration {
    text-align: center;
  }
  
  .button {
    border: none;
    outline: none;
    font-size: 11px;
    font-family: "Quicksand", sans-serif;
    color: #f44336;
    padding: 3px 10px;
    border-radius: 8px;
    cursor: pointer;
    border: 1px solid #f44336;
    background-color: transparent;
    margin-right: 5px;
  }
  
  .button:active {
    border: 1px solid blue;
  }