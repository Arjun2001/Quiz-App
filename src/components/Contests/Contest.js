import React, { useState, useEffect } from 'react'
import axios from 'axios'
import Swal from 'sweetalert2'
import Button from 'react-bootstrap/Button'
import './Contest.css'
import {useHistory} from "react-router-dom";

import Navbar from '../Navbar/Navbar'


const Table = () => {
    const [contest, setContest] = useState([]);
    const history = useHistory();

    useEffect(() => {
        getData()
    }, [])

    const getData = async () => {
        try {
            await axios ({
                method:'post',
                url: "http://35.225.238.45:5000/api/contests",
                headers: {
                    "Authorization":`Bearer ${localStorage.getItem('Token')}`,
                    "Content-Type": "application/json"
                },
                data: {
                    code:window.location.pathname.substring(9)
                }
            }).then(res => {
                setContest(res.data)
            })
        }catch (err) {
            console.log(err);
        }
    }

    const convertTime = (START) => {
        let time = new  Date(START);
        time = time.toISOString();
        time = (new Date(time).toLocaleString(undefined, {timeZone: 'Asia/Kolkata'}))
        return time;
    }

    const EditQuiz = (id) => {
        history.push(`/quiz/${id}`)
    }

    const Details = (id) => {
        history.push(`/stats/${id}`)
    }

    const details_p = (id) => {
        history.push(`/details/${id}`)
    }
    
    const Average = (id) => {
        history.push(`/performance/${id}`)
    }
                // hast ot chenges when linked with average

    const joinQuiz = (e) => {
        try {
            axios ({
                method:'post',
                url: "http://35.225.238.45:5000/api/contest_details",
                headers: {
                    "Authorization":`Bearer ${localStorage.getItem('Token')}`,
                    "Content-Type": "application/json"
                  },
                  data: {
                    id:e
                }
              }).then(res => {
                let time = new Date();
                time = time.toISOString();
                let st_time = res.data.data[0].START
                let end_time = res.data.data[0].END
                if ( st_time <= time && end_time >= time ) {
                    Swal.fire({
                        title: 'Enter the passcode',
                        html: `<input type="password" id="pin" class="swal2-input">`,
                        confirmButtonText: 'Join',
                        focusConfirm: false,
                        preConfirm: () => {
                          const pin = Swal.getPopup().querySelector('#pin').value
                          if (!pin) {
                            Swal.showValidationMessage(`Please enter the passcode`)
                          }
                          return pin
                        }
                      }).then((result) => {
                          console.log(result , res.data.data[0].passcode)
                        if (result.value === res.data.data[0].passcode) {
                            Swal.fire({
                            icon: 'success',
                            text: "Passcode is correct"
                            }).then(results => {
                                history.push(`/join/${e}`)
                            })
                        } else {
                            Swal.fire({
                            icon: 'error',
                            text: "Passcode is incorrect"
                            })
                        }
                    })
                } else {
                    Swal.fire({
                        icon: 'info',
                        text: "Contest has not started yet"
                        })
                }
              })
            }catch (err) {
                console.log(err);
            }
    }


    const renderHeader = () => {
        let headerElement = ['ID', 'Course Code', 'Name', 'Start Time', 'End Time','action']

        return headerElement.map((key, index) => {
            return <th key={index}>{key.toUpperCase()}</th>
        })
    }

    const renderPresent = () => {
        return contest && contest.map(({CODE,END,START,NAME,ID},index) => {
            let now = new Date();
            now = now.toISOString();
            now = (new Date(now).toLocaleString(undefined, {timeZone: 'Asia/Kolkata'}))
            
            if ((now <= START) && (now > END)) {
                return (
                    <tr key={ID}>
                        <td>{ID}</td>
                        <td>{CODE}</td>
                        <td>{NAME}</td>
                        <td>{convertTime(START)}</td>
                        <td>{convertTime(END)}</td>
                        <td className='opration'>
                        {localStorage.getItem('Role') === "Admin" ? <button className='button' onClick={() => {EditQuiz(ID)}}>Edit</button>:<button className='button' onClick={() => {joinQuiz(ID)}}>Join</button>}
                        </td>
                    </tr>
                )
            }
        })
    }

    const renderPast = () => {
        return contest && contest.map(({CODE,END,START,NAME,ID},index) => {
            let now = new Date();
            now = now.toISOString();
            now = (new Date(now).toLocaleString(undefined, {timeZone: 'Asia/Kolkata'}))
            if (now > START) {
                return (
                    <tr key={ID}>
                        <td>{ID}</td>
                        <td>{CODE}</td>
                        <td>{NAME}</td>
                        <td>{convertTime(START)}</td>
                        <td>{convertTime(END)}</td>
                        <td className='opration'>
                        {localStorage.getItem('Role') === "Admin" ? <><button className='button' onClick={() => {EditQuiz(ID)}}>Edit</button><button className='button' onClick={() => {details_p(ID)}}>Details</button></>:<div></div>}
                            {localStorage.getItem('Role') === "Student" ?<button className='button' onClick={() => {Details(ID)}}>Details</button> : <button className='button' onClick={() => {Average(ID)}}>Average</button>}
                        </td>
                    </tr>
                )
            }
        })
    }

    const renderFuture = () => {
        return contest && contest.map(({CODE,END,START,NAME,ID},index) => {
            let now = new Date();
            now = now.toISOString();
            now = (new Date(now).toLocaleString(undefined, {timeZone: 'Asia/Kolkata'}))
            if (now < START) {
                return (
                    <tr key={ID}>
                        <td>{ID}</td>
                        <td>{CODE}</td>
                        <td>{NAME}</td>
                        <td>{convertTime(START)}</td>
                        <td>{convertTime(END)}</td>
                        <td className='opration'>
                            {localStorage.getItem('Role') === "Admin" ? <button className='button' onClick={() => {EditQuiz(ID)}}>Edit</button>:<div></div>}
                        </td>
                    </tr>
                )
            }
        })
    }

    const create = () => {
        Swal.fire({
            title: 'Create Contest',
            html: `<label for="name">Contest Name</label><input type="text" id="name" class="swal2-input" placeholder="Contest Name">
            <label for="start">Start Time</label><input type="datetime-local" id="start" class="swal2-input" placeholder="Start Time">
            <label for="end">End Time</label><input type="datetime-local" id="end" class="swal2-input" placeholder="End Time">
            <label for="passcode">Passcode</label><input type="text" id="passcode" class="swal2-input" placeholder="Passcode">`,
            confirmButtonText: 'Submit',
            focusConfirm: false,
            preConfirm: () => {
              const name = Swal.getPopup().querySelector('#name').value
              const start = Swal.getPopup().querySelector('#start').value
              const end = Swal.getPopup().querySelector('#end').value
              const passcode = Swal.getPopup().querySelector('#passcode').value
              if (!name || !start || !end || !passcode) {
                Swal.showValidationMessage(`Please enter all the details`)
              }
              return { name: name, start: start, end: end , passcode: passcode}
            }
          }).then((result) => {
            try {
                axios ({
                    method:'post',
                    url: "http://35.225.238.45:5000/api/create_contest",
                    headers: {
                        "Authorization":`Bearer ${localStorage.getItem('Token')}`,
                        "Content-Type": "application/json"
                      },
                      data: {
                        code:window.location.pathname.substring(9),
                        name:result.value.name,
                        start:result.value.start,
                        end:result.value.end,
                        passcode:result.value.passcode
                    }
                }).then(res => {
                    getData();
                    Swal.fire({
                        icon: 'success',
                        text: "Contest Created successfully"
                    })
                })
            }catch (err) {
                console.log(err);
            }
          })
    }

    return (
        <>
        <Navbar />
        <div style={{display:"flex", flexDirection:'column', margin:"50px 75px"}}>
            {localStorage.getItem('Role') === "Admin" ?<Button variant="primary" size="lg" block onClick={create}>
                CREATE CONTEST
            </Button> : <div></div>}
            <h4 id='title'>Present Contests</h4>
            <table id='employee'>
                <thead>
                    <tr>{renderHeader()}</tr>
                </thead>
                <tbody>
                    {renderPresent()}
                </tbody>
            </table>
            <h4 id='title'>Future Contests</h4>
            <table id='employee'>
                <thead>
                    <tr>{renderHeader()}</tr>
                </thead>
                <tbody>
                    {renderFuture()}
                </tbody>
            </table>
            <h4 id='title'>Past Contests</h4>
            <table id='employee'>
                <thead>
                    <tr>{renderHeader()}</tr>
                </thead>
                <tbody>
                    {renderPast()}
                </tbody>
            </table>
        </div>
        </>
    )
}


export default Table