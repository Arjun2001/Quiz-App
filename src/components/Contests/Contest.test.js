import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Contest from './Contest';

describe('Contest component', () => {
    it('div rendering', () => {
        const wrapper = shallow(<Contest />);
        const text = "CREATE CONTEST"
        wrapper.find('button');
        expect(wrapper.contains(text)).toEqual(true);
    });
    it('heading rendering', () => {
        const wrapper = shallow(<Contest />);
        var text = "Present Contests"
        wrapper.find('#title');
        expect(wrapper.contains(text)).toEqual(true);
        text = "Future Contests";
        expect(wrapper.contains(text)).toEqual(true);
        text = "Past Contests"
        expect(wrapper.contains(text)).toEqual(true);
    });

    it('matches the snapshot', () => {
        const tree = renderer.create(<Contest />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});