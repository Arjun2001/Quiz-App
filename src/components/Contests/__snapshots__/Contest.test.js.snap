// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contest component matches the snapshot 1`] = `
Array [
  <nav
    className="navbar-bg navbar navbar-expand navbar-dark"
  >
    <a
      className="navbar-brand"
      href="/home"
    >
      <img
        src="https://cdn.pixabay.com/photo/2018/08/31/11/17/quiz-3644414_960_720.png"
        style={
          Object {
            "width": "60px",
          }
        }
      />
    </a>
    <button
      aria-controls="responsive-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="responsive-navbar-nav"
    >
      <div
        className="mr-auto navbar-nav"
        onKeyDown={[Function]}
      />
      <div
        className="navbar-nav"
        onKeyDown={[Function]}
      >
        <span
          className="navbar-text"
        >
          Signed in as: 
          <strong
            style={
              Object {
                "color": "white",
              }
            }
          />
        </span>
        <div
          className="nav-item dropdown"
        >
          <a
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle nav-link"
            data-rb-event-key={null}
            href="#"
            id="basic-nav-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
          />
        </div>
      </div>
    </div>
  </nav>,
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "margin": "50px 75px",
      }
    }
  >
    <button
      className="btn btn-primary btn-block btn-lg"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      CREATE CONTEST
    </button>
    <h4
      id="title"
    >
      Present Contests
    </h4>
    <table
      id="employee"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            COURSE CODE
          </th>
          <th>
            NAME
          </th>
          <th>
            START TIME
          </th>
          <th>
            END TIME
          </th>
          <th>
            ACTION
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <h4
      id="title"
    >
      Future Contests
    </h4>
    <table
      id="employee"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            COURSE CODE
          </th>
          <th>
            NAME
          </th>
          <th>
            START TIME
          </th>
          <th>
            END TIME
          </th>
          <th>
            ACTION
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <h4
      id="title"
    >
      Past Contests
    </h4>
    <table
      id="employee"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            COURSE CODE
          </th>
          <th>
            NAME
          </th>
          <th>
            START TIME
          </th>
          <th>
            END TIME
          </th>
          <th>
            ACTION
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>,
]
`;
